### ISSUE TITLE
Locator Devices

### ISSUE THREAD

**Issue by j4james** [hackerb9/vt340test#35]
I saw on the home page README that you might be trying to get hold of a locator device to test with, but in case you can't find one, you might be interested in a project I came across on _Hackaday_ that shows how to build an adapter for connecting a PS/2 mouse to the VT340.

https://hackaday.io/project/19576-dec-mouse-adapter

I know absolutely nothing about hardware hacking, so I don't know how easy that would be to build, but I figured it might be something you would understand.

Also, I noticed that the links in the [Locator Devices](https://github.com/hackerb9/vt340test/blob/087ec9d409c0a3481419ecd0939447020152055d/README.md#locator-devices) section aren't rendering correctly on GitHub. I think the way it's supposed to work if you're using the footnote link format is you'll use a reference in the text like this:
```
the common protocol they spoke in the [VCB02 manual], particularly the appendices...
```

And then define the link for that reference in a footnote like this:
```
[VCB02 manual]: docs/kindred/EK-104AA-TM-001_VCB02_Video_Subsystem_Feb87.pdf
```

If you want to test stuff like this before checking it in, what I do is paste my markdown into a comment or issue on GitHub (obviously without submitting it) and just see what it looks like in the preview.

**Comment by j4james** [comment]
And while we're on the subject of locator protocols, there was something else I wanted to discuss with you. I've seen a couple of places where you've mentioned the capabilities of DECterm and the VWS VT200 emulator, and I was curious whether you actually have access to a system with either of those terminals?

The reason I ask is because both of them are supposed to support the [DEC Text Locator extension](https://web.archive.org/web/20110103061348/http://web.eecs.utk.edu/~shuford/terminal/dec_vt_mouse.html), which is something I wanted to implement in Windows Terminal. The documentation can be a bit ambiguous (as is often the case), so I was hoping to find a reasonable reference implementation I could test against. And if the VT340 doesn't have that extension, then a terminal emulator written by DEC would be the next best thing.

**Comment by Asmodai** [comment]
Not sure if you got to test DECterm, but I have a couple of variants here.

I've tested DEC Locator support with DECterm on both ULTRIX 4.5 VAX and Digital UNIX 4.0F.
I used vttest-20251205 with the DEC Locator mouse tests.

Both terminals provide locator support with character cell and pixel modes.

**Comment by hackerb9** [comment]
Thank you, Paul. I recently got an actual locator for my VT340 — DEC VSXXX-AA — but haven't had a chance to test it yet.

@j4james : did I already share the link with you for how to run decterm in emulation?

**Comment by j4james** [comment]
> I used vttest-20251205 with the DEC Locator mouse tests.

@Asmodai Assuming you ran the "2. Repeated" test, can I ask what you saw being reported when clicking and releasing a mouse button? Because in some terminals (including one of DEC's Windows terminal emulators), I saw something like this:

1. When the left button is clicked:
   * `left button down - right` (when pressed)
   * `left button up -  no buttons down` (when released)
2. When the middle button is clicked:
   * `middle button down - middle` (when pressed)
   * `middle button up - no buttons down` (when released)
3. When the right button is clicked:
   * `right button down - left` (when pressed)
   * `right button up - no buttons down` (when released)

That left-right, right-left mismatch is assumedly not what vttest was expecting, but I think it's actually correct. So I'm curious whether DECterm gave you the same results.

**Comment by j4james** [comment]
> did I already share the link with you for how to run decterm in emulation?

@hackerb9 I don't think you did. Although I have previously seen instructions on setting up an emulation of OpenVMS, which I assumed would enabled me to run DECterm, but that seemed like it was way beyond the capabilities of my little laptop. If there's an easier way to test DECterm I'd love to know.

**Comment by Asmodai** [comment]
I believe there's ULTRIX 4.5 images for opensimh (MicroVAX 3900 targets) that will let you use UWS via a remote X server.
Another quick way would be GXEmul with ULTRIX 4.5 MIPS.  I believe both have images on gunkies or similar and both should run on fairly low-powered hardware.

I ran the entire suite of DEC locator tests.  For the "repeated" test, I get similar results.  However, you can chord mouse buttons.

left button: `left button down - left`
and right button: `right button down - right, left`
release right: `right button up - left`
then release left: `left button up - no buttons down`

Out of interest, I tried this with one of the better VT340 emulators I have -- Reflection 4 (albeit older version for Windows 3), and it failed all the locator tests.

**Comment by j4james** [comment]
> left button: `left button down - left`

@Asmodai Thanks for the info, but that unfortunately doesn't match what I was seeing. And that means that DEC wasn't implementing the protocol consistently on all their terminal emulators, so it's harder to argue that any one interpretation is correct.

> Out of interest, I tried this with one of the better VT340 emulators I have -- Reflection 4 (albeit older version for Windows 3), and it failed all the locator tests.

That's understandable, though. Reflection is what I'd call a "real" terminal emulator, so they tend to stick to the features that were present on the original hardware terminals, and there weren't any that supported the text locator mode as far as I'm aware.

Reflection does support the VT340 locator port extension, though, so you can at least use ReGIS graphics input commands to query the mouse.

**Comment by Asmodai** [comment]
Out of interest, which terminal emulator was giving inconsistent results?

**Comment by j4james** [comment]
I think there were a couple that interpreted the button state differently to what vttest expected, but the one I was most concerned about was VTStar, since that was produced by DEC. If DECterm had matched VTStar, then I'd be more confident that vttest was wrong. But as things stand, there's no obviously right answer.

**Comment by Asmodai** [comment]
Here's some output from VTstar VT320e v1.3's datascope when I run the locator test.

<img width="872" height="454" alt="Image" src="https://github.com/user-attachments/assets/728747cd-79b5-4ec3-8821-f1345030b4af" />

Column on the left is data sent by the host, and the column on the right is the data sent by the terminal.

We can clearly see it emit `CSI 2 ; 1 ; 13 ; 34 & w`:
`Pe` = `2` -- Left button down
`Pb` = `1` -- Right

Then we next see `CSI 3 ; 0 ; 13 ; 34 & w`:
`Pe` = `3` -- Left button up
`Pb` = `0` -- No buttons down

So the initial button down event appears to have an incorrect button code.

I decoded these via EL-00070-13 -- "DEC STD 070-13 Video Systems Reference Manual--Text Locator Extension" (1991), Locator Report [DECLRP], page 6.

This can be viewed at page 1021 of https://www.bitsavers.org/pdf/dec/standards/EL-SM070-00_DEC_STD_070_Video_Systems_Reference_Manual_Dec91.pdf

The values for `Pe` are described as thus:
| `Pe` event  code | Meaning                                                                                                  |
|:----------------:|----------------------------------------------------------------------------------------------------------|
| `0`              | Request; the terminal received an explicit request for a locator report, but the locator is unavailable. |
| `1`              | Request; the terminal received an explicit request for a locator report.                                 |
| `2`              | Left button down.                                                                                        |
| `3`              | Left button up.                                                                                          |
| `4`              | Middle button down.                                                                                      |
| `5`              | Middle button up.                                                                                        |
| `6`              | Right button down.                                                                                       |
| `7`              | Right button up.                                                                                         |
| `8`              | Fourth button down.                                                                                      |
| `9`              | Fourth button up.                                                                                        |
| `10`             | Locator outside filter rectangle.                                                                        |

The values for `Pb` are described as thus:
| `Pb` button code | Meaning          |
|:----------------:|------------------|
| `0`              | No buttons down. |
| `1`              | Right.           |
| `2`              | Middle.          |
| `4`              | Left.            |
| `8`              | Fourth           |

The important part is:
> `Pb` is the button code, ASCII decimal 0 - 15, indicating which, if any, of up to 4 buttons are down.  The state of up to 4 buttons on a locator corresponds to the low 4 bits of the decimal value; a `1` means a button is depressed.

So, for a left button down event, I'd expect `Pb` to be `4`.

Now interestingly, I wonder if VTstar shares common source code with DECterm.  If so, there's a nice comment in OSF/1 DECterm sources (in the locator handling code) about button codes being inverted due to the way DECwindows reports X11 mouse events.

Does this tally with what you're seeing too?  I'd be curious to throw more terminal emulators at this.  Mostly because terminal emulators are interesting, but secondly because I wonder how many copied what VTstar did.

Edited to give right product name... _VT320e_ not _VT230e_

Further edit:
I'll put this here so the Internet never forgets: https://www.digiater.nl/openvms/freeware/v70/vtstar/kit/
This is the terminal emulator that was part of the MultiClient Desktop for the Intel-based DEC Multia workstation.  I've tried searching for the full thing, but no joy.  Whoever archived the terminal emulator did us a huge favour.

**Comment by davidrg** [comment]
> I wonder if VTstar shares common source code with DECterm

I've wondered this too - surely DEC wouldn't write _three_ (or more? what was on the VT LAN40?) separate terminal emulators from scratch. But VTstar, AFAIK, doesn't do ReGIS or Sixel. Haven't checked ruled lines, but I assume thats DECterm only too (and buggy in 5.1B at least). And neither seem to support the VWS terminal emulators clipboard stuff.

**Comment by j4james** [comment]
@Asmodai I'm seeing the same thing you're seeing, but the reason I think VTStar is correct (and DECterm is wrong), is because I believe the _Pb_ button code documentation was wrong.

The locator protocol is designed to support a range of devices with up to 4 buttons, which they sometimes refer to generically as B1, B2, B3, and B4. You can see the mappings of those button numbers in various places, such as the [`DECLKD`](https://j4james.github.io/vtdocs/EL-00070-13.html#DECLKD) documentation in STD 070.

* B1 = left = barrel button
* B2 = middle = tip button
* B3 = right
* B4 (tablet cursor only)

And when you look at the documentation for the _Pe_ parameter of [`DECLRP`](https://j4james.github.io/vtdocs/EL-00070-13.html#DECLRP), it follows the same order:

* 2/3 Left button down/up (i.e. B1)
* 4/5 Middle button down/up (i.e. B2)
* 6/7 Right button down/up (i.e. B3)
* 8/9 Fourth button down/up (i.e. B4)

So when you get to the _Pb_ parameter, and it says "the state of up to 4 buttons on a locator corresponds to the low 4 bits of the decimal value", the natural assumption would be.

* 1 = B1 (i.e. left button)
* 2 = B2 (i.e. middle button)
* 4 = B3 (i.e. right button)
* 8 = B4 (i.e. fourth button)

But the documentation (the table you quoted previously) has the right and left buttons swapped, implying it's actually B3, B2, B1, B4. When you put it like that, it seems obviously wrong. And if you were using a stylus, which only has two "buttons" (barrel and tip), this implies the button state would be represented by values 2 and 4, rather than 1 and 2, which again seems wrong.

As one further bit of "confirmation", the first edition of the VT330/VT340 graphics programmer reference manual made the exact same mistake in the summary documentation for the `DECLBD` (aka `DECLKD`) sequence, listing the button order like this (in table 15-3 on page 254):

| Ky | Button | Device                        |
|----|--------|-------------------------------|
| 1  | Right  | Mouse, puck, or stylus barrel |
| 2  | Middle | Mouse, puck, or stylus tip    |
| 3  | Left   | Mouse or puck                 |
| 4  | Fourth | Puck only                     |

And note that the stylus assignments are correct - it's just the `Right` and `Left` button labelling that is wrong. And it was only that summary table that got it wrong - they had the correct labels earlier in the manual (page 251).

> I'd be curious to throw more terminal emulators at this. Mostly because terminal emulators are interesting, but secondly because I wonder how many copied what VTstar did.

DEC also had a VWS VT200 emulator emulator, which I believe supported the text locator mode. So if anyone has access to a VMS workstation, they should be able to test that. Assuming this is actually a separate product to DECterm, and not just another name they used.

**Comment by hackerb9** [comment]
> DEC also had a VWS VT200 emulator emulator, which I believe supported the text locator mode. So if anyone has access to a VMS workstation, they should be able to test that. Assuming this is actually a separate product to DECterm, and not just another name they used.

Oops, I meant to say DEC VWS VT200/ReGIS terminal, not DECterm. As @Asmodai said, it is possible with simh and does not require a powerful CPU to emulate. My main problem was  finding a VMS/VWS disk image that worked. I found one on 9track.net that boots, but only with an old, buggy version of simh (not OpenSIMH). It is good enough for my testing. However, the main flaw in it is that the mouse is almost unusably erratic. 

I will find my notes on how to get it running and link to them from here.

**Comment by Asmodai** [comment]
The VWS emulator requires the VMS Workstation Software, and it eventually became DECterm, with the sources for OSF/1 containing code for VMS, ULTRIX, and VXT, with direct reference to the VAXstation II.

Recent VMS will have X11, and thus DECterm.

It is possible to build a VAX simulation with simh that includes framebuffer support: https://www.9track.net/simh/video/

I haven't done any experimentation with this myself, but I do plan to.  Perhaps I can experiment with it this weekend and get an answer from the VWS terminal emulator.

The documentation does match DEC STD 070, however.

From "VMS Workstation Software User's Guide" (for VWS 4.1, AA-EZ24D-TE):
<img width="729" height="1126" alt="Image" src="https://github.com/user-attachments/assets/b655520b-a363-4899-ba61-460cfe8826b2" />
Granted, this could be a copy/paste job from DEC STD 070 and proves nothing.

As a bit of extra work, I explored the way the windowing system that the VWS emulator uses: UIS.

The way the VWS emulator handled buttons would most likely be via `UIS$GET_BUTTONS`.  Here's the relevant portion of the "VMS Workstation Software 4.1 Graphics Programming Guide" (AA-GI10C-TE):
<img width="747" height="958" alt="Image" src="https://github.com/user-attachments/assets/18238017-042e-4549-95de-08805ed0f56f" />

I did look up the differences between UIS and X11 to see if I could find any interesting clues, but the documentation only really explains the differences between the event systems from a high-level standpoint, and refuses to get into the weeds on the differences between buttons and the like.

From other functions in the documentation, I was able to ascertain some of the definition meanings:
left button = `UISDC$C_POINTER_BUTTON_1`
center button = `UISDC$C_POINTER_BUTTON_2`
right button = `UISDC$C_POINTER_BUTTON_3`
I guess `UISDC$C_POINTER_BUTTON_4` is for a fourth button that some devices might have, the documentation isn't explicit.

There's no explaination on whether `UIDC$C_POINTER_BUTTON_` correlates with `UIS$M_POINTER_BUTTON_` though.  I guess I'll need to extract the headers.

I don't want to assume that these are defined as button 1 = `0x01`, button 2 = `0x02` and so on but this does add weight to your position.

I think the best path forward would be to experiment with VWS.  Looks like the QVSS hackery on 9track.net is using VMS 5.4.  I've grabbed a copy of that from bitsavers, and I'll have a play with the QVSS stuff.

**Comment by hackerb9** [comment]
Yes, I am using 9track.net's old simh with framebuffer support. The VMS/VWS disk image I used is here https://bit.ly/smalltalk-80-vax .

**Comment by Asmodai** [comment]
@hackerb9 I get much better mouse support with that image by editing the ini file;
Nuke the `SET QVSS CAPTURE` line and add `SET VC CAPTURE`.

**Comment by j4james** [comment]
> The documentation does match DEC STD 070, however.

Yeah, I've seen four different versions of the locator documentation, and the wording varies quite a bit between them, but the listed values for the _Pb_ parameter are pretty much identical in every case. I realise that's a strong argument against VTStar being correct, but it's not impossible that they've all just cut and paste that bit of the documentation from the original source.

For the record, these are the references I've seen (ordered from oldest to newest I think):
1. VMS Workstation Software Users Guide (AA-EZ24D-TE)
2. [Locator Input Model for ANSI Terminals] (sixth revision)
3. VT1000 Users Guide (EK-V1000-UG-002)
4. Chapter 13 of DEC STD 070 (EL-00070-13)

[Locator Input Model for ANSI Terminals]: https://espterm.github.io/docs/Locator%20Input%20Model%20for%20ANSI%20Terminals%20(sixth%20revision).html

**Comment by hackerb9** [comment]
> Nuke the `SET QVSS CAPTURE` line and add `SET VC CAPTURE`.

Thanks, but that didn't help for my setup. It is something to do with how simh reads the input from Linux as the mouse behaves properly if I switch virtual terminals and switch back.

**Comment by hackerb9** [comment]
Here are my notes for using simh to run VMS/VWS and open a VT200/ReGIS terminal. 
  https://gist.github.com/hackerb9/929756faa2d459a5cea8431b3a39ff7f

**Comment by Asmodai** [comment]
<img width="1044" height="908" alt="Image" src="https://github.com/user-attachments/assets/d4955df8-30ae-4ace-8ab4-566ac428a567" />

I think we have an answer.

To set terminal up, as `SYSTEM` account:

```dcl
DEFINE UIS$VT_ENABLE_LOCATOR TRUE /SYSTEM
```

This allows the emulator to respond with the ANSI locator device attribute.

To prime the terminal for my shenanigans, I use the following DCL in a file called `setterm.com`
```dcl
$ ESC[0,8] = 27
$
$! Configure terminal.
$ SET TERMINAL/PASTHRU/NOECHO/ESCAPE
$
$! Clear screen.
$ WRITE SYS$OUTPUT ESC + "c"
$
$! Configure locator.
$ WRITE SYS$OUTPUT ESC + "[0)u"
$ WRITE SYS$OUTPUT ESC + "[13;{"
$ WRITE SYS$OUTPUT ESC + "[1;2'z"
$ WRITE SYS$OUTPUT ESC + "[1'|"
```

I use the following gnarly C for the test.  I can't type `<` or `>` which I find really hilarious and entertaining... so... K&R it is.
```c
int main()
{
  int c = 0;

  extern int printf();
  extern int fflush();
  extern int getchar();

  extern char *stdout;

  while ((c = getchar()) != 0) {
    switch (c) {
    case 'q':
      fflush(stdout);
      exit(1);
      break;

    case '\n':   printf("\r\n");     break;
    case '\033': printf("\r\nESC "); break;
    case '[':    printf("[ ");       break;    
    default:     printf("%c ", c);
    }
  }
  
  return 1;
}
```

To reset the terminal when done, the following can be used:
```dcl
SET TERMINAL/NOPASTHU/ECHO/NOESCAPE
```

I have a DCL file I use to compile and link the C, I call it `mcsi.com`
```dcl
$! Make CSI tool.
$ CC csi.c
$ LINK csi,sys$library:vaxcrtl/LIBRARY
```

The whole thing is executed as this:
```dcl
@mcsi
@setterm
run csi
set terminal/nopasthru/echo/noescape
```

**Comment by Asmodai** [comment]
@hackerb9 Thanks for linking that image, by the way.  I have a full set of VMS 5.5 + supplemental somewhere, so now I have no excuse for setting up a VWS system.  Ahh, if only I had infinite time.

**Comment by j4james** [comment]
@Asmodai I think that clinches it. As much as I dislike the result, VTStar is clearly in the minority here, so I'll have to accept the DECterm implementation as the correct one. Which is probably best for interop anyway, if most terminals follow the documentation exactly as written. Thanks for testing this.

**Comment by Asmodai** [comment]
You're welcome, it was fun!

I still have questions, though.  I wonder why VT320e has that behaviour.  If it's in 1.4 and in 1.3, then it's probably been there since the initial internal builds.

I suppose I'll never find out unless I find the source code.  I suspect an engineer imported parts of DECterm without reading the comments about the inversion of buttons.

I'd also like to hope that the DEC Multia desktop (either in "MultiaClient Desktop" or "Multia Enterprise" form) is sitting on some unknown FTP server somewhere waiting to be discovered.

I'd also be curious to see how VWS's VT200 emulator compares to the one on the DECpro.

**Comment by hackerb9** [comment]
Thank you, @Asmodai.  I hope to run into  you again as your knowledge of VAX/VMS arcana was extremely helpful.

[I'm leaving this bug open as a reminder to myself to update the locator documentation. ]

**Comment by j4james** [comment]
I should add that there are a bunch more edge cases that would be good to test if anyone is enthusiastic enough. Some aspects of the protocol just aren't documented at all, and some things are documented but with conflicting requirements depending on which manual you're reading. It's been a few years since I looked at this, though, so I need to dig up my old notes and see if I've got a usable test script that I can share.

**Comment by j4james** [comment]
I'm not sure if this is a suitable language for testing on these systems, by I've created a little python script that attempts to record some of the edge cases for the locator report here: [declrp.py](https://gist.github.com/j4james/6b3a8fea1b8df8f530c65f5608577ea1).

On startup it performs a hard reset and then gives the terminal 5 seconds to recover. You'll then see some instructions asking you to position the mouse cursor in the centre of the screen and to press ENTER. The rest should be automatic. It runs through a series of 22 [`DECRQLP`](https://j4james.github.io/vtdocs/EL-00070-13.html#DECRQLP) queries with a half-second pause between each, so shouldn't take much more than 11 seconds. The results are written to a log file named `declrp.log`.

This is just testing how the `DECLRP` report adjusts to various terminal states, for example:
- double width lines
- 132-column mode (assuming the terminal uses a narrow-width font for that)
- coordinates inside and outside margins with a relative origin mode
- active vs visible pages (assuming multiple pages are supported)

This is not urgent, but if anyone wants to give this a go on DECterm and/or the VWS emulator, I'd be interested to see what results they produce. VTStar's implementation is quite buggy, so I wouldn't be confident relying on that as a reference implementation.

**Comment by Asmodai** [comment]
As neither ULTRIX or Digital UNIX are capable of running Python 3, I ran the tests on a FreeBSD 15 machine accessed via telnet.

I can't easily test VWS just yet as... well, VMS 5 is ancient now and I don't have any TCP/IP stacks for it.  Native networking is done via DECnet, and I have yet to finish setting up my DECnet segment.  As its also running via simh on WIndows I can't easily set up any LAT hacks either.  So, VWS will have to wait.

DECterm on Digital UNIX 4.0:
```
Page Size             (24, 80)
After Hard Reset      '\x1b[0&w'
Cell Request          '\x1b[1;0;12;40&w'
Pixel Request         '\x1b[1;0;149;276&w'
DECDWL Cell Request   '\x1b[1;0;12;20&w'
DECDWL Pixel Request  '\x1b[1;0;149;276&w'
DECCOLM Cell Request  '\x1b[1;0;13;56&w'
DECCOLM Pixel Request '\x1b[1;0;148;276&w'
Cell In Margins       '\x1b[1;0;3;40&w'
Pixel In Margins      '\x1b[1;0;149;276&w'
Cell Out Left         '\x1b[1;0;3;40&w'
Pixel Out Left        '\x1b[1;0;149;276&w'
Cell Out Right        '\x1b[1;0;3;40&w'
Pixel Out Right       '\x1b[1;0;149;276&w'
Cell Out Above        '\x1b[0&w'
Pixel Out Above       '\x1b[1;0;149;276&w'
Cell Out Above        '\x1b[0&w'
Pixel Out Above       '\x1b[1;0;149;276&w'
Page 2 Visible        '\x1b[1;0;12;40&w'
Page 2 Not Visible    '\x1b[1;0;12;40&w'
Disabled              '\x1b[0&w'
One Shot On           '\x1b[1;0;12;40&w'
One Shot Off          '\x1b[0&w'
```

DECterm on ULTRIX 4.5 VAX:
```
Page Size             (24, 80)
After Hard Reset      '\x1b[0&w'
Cell Request          '\x1b[1;0;12;40&w'
Pixel Request         '\x1b[1;0;150;277&w'
DECDWL Cell Request   '\x1b[1;0;12;20&w'
DECDWL Pixel Request  '\x1b[1;0;150;277&w'
DECCOLM Cell Request  '\x1b[1;0;13;56&w'
DECCOLM Pixel Request '\x1b[1;0;149;277&w'
Cell In Margins       '\x1b[1;0;3;40&w'
Pixel In Margins      '\x1b[1;0;150;277&w'
Cell Out Left         '\x1b[1;0;3;40&w'
Pixel Out Left        '\x1b[1;0;150;277&w'
Cell Out Right        '\x1b[1;0;3;40&w'
Pixel Out Right       '\x1b[1;0;150;277&w'
Cell Out Above        '\x1b[0&w'
Pixel Out Above       '\x1b[1;0;150;277&w'
Cell Out Above        '\x1b[0&w'
Pixel Out Above       '\x1b[1;0;150;277&w'
Page 2 Visible        '\x1b[1;0;12;40&w'
Page 2 Not Visible    '\x1b[1;0;12;40&w'
Disabled              '\x1b[0&w'
One Shot On           '\x1b[1;0;12;40&w'
One Shot Off          '\x1b[0&w'
```

**Comment by Asmodai** [comment]
Oh, as a quick note... neither DECterm or the VWS terminal emulator support multiple pages.

**Comment by j4james** [comment]
@Asmodai Thank you so much. These results are very close to what I was expecting to see. Some points of interest:

1. On a double width line, the center column is reported as 20 rather than 40. This is what I was expecting, but most modern terminal emulators do _not_ handle this correctly.
2. 132-column mode also affects the coordinates in a similar way, although I was expecting something closer to 66 for the column number rather than 56. I'm guessing that might be affected by the font used for that mode, because the row coordinate also appears to be off by 1.
3. Cell coordinates are affected by the margins, but pixel coordinates are not. I wasn't sure about the latter, but this behaviour seems reasonable to me. I don't think any modern terminal emulators handle margins correctly though.
4. When outside the margins, it reports the locator as unavailable (`\x1b[0&w`) rather than an undefined position (`\x1b[1;0&w`). This aligns with the older VWS documentation, but the more recent STD 070 documentation implies the latter response would be more appropriate. Both options seem reasonable to me.

> neither DECterm or the VWS terminal emulator support multiple pages.

Thanks for confirming that. I thought that might be the case, but I threw in the page test on the off chance it might work. Same goes for the horizontal margins actually. I'm assuming DECterm doesn't support that either.

**Comment by Asmodai** [comment]
>     2. 132-column mode also affects the coordinates in a similar way, although I was expecting something closer to 66 for the column number rather than 56. I'm guessing that might be affected by the font used for that mode, because the row coordinate also appears to be off by 1.

Possibly.  By default it'll resize the window and change to a condensed font.  That can be disabled, so I'll have a poke at that tomorrow.


> Thanks for confirming that. I thought that might be the case, but I threw in the page test on the off chance it might work. Same goes for the horizontal margins actually. I'm assuming DECterm doesn't support that either.

I can't see any mention of horizontal margins in the documentation or the source code, so I assume your assumption is the right one here :)

**Comment by davidrg** [comment]
> Oh, as a quick note... neither DECterm or the VWS terminal emulator support multiple pages.

Interesting. I'm pretty sure when I was testing it on Tru64 5.1B DECterm did support multiple pages - I wonder when that was added, or if I was just imaginging things. I'll have to do another test some time.

This makes me wonder more what the behaviour of Ruled Lines is like on earlier versions of DECterm. When was it added and was it always kind of broken, or did some change in some later version of DECterm introduce bugs?

**Comment by Asmodai** [comment]
```c
s_do_locator_report(w, event, button, row, column, page)
DECtermWidget w;
int event, button, row, column, page;
{
        wvtp ld = w_to_ld( w );
        char report[100];
        Boolean disabled = ( _cld wvt$l_vt200_common_flags &
          vtc1_m_locator_report_mode );

        s_doing_report(w);
        if ( row < 0 || event == 0 || disabled )
                sprintf(report, "\2330&w");  /* locator unavailable */
        else
                sprintf(report, "\233%d;%d;%d;%d&w",
                        event, button, row, column);
                /* omit the page parameter because it is always 1 */
        i_report_data(w, report, strlen(report));
}
```

You might be right.  DECterm appears to support up to 6 pages, just not for the locator report.  Not sure if this was changed in later releases of Digital UNIX / Tru64.

**Comment by hackerb9** [comment]
> I can't easily test VWS just yet as... well, VMS 5 is ancient now and I don't have any TCP/IP stacks for it. Native networking is done via DECnet, and I have yet to finish setting up my DECnet segment. As its also running via simh on WIndows I can't easily set up any LAT hacks either. So, VWS will have to wait.

I've done some more poking at VWS in SIMH and found a [slightly better way](https://gunkies.org/wiki/Installing_MicroVMS_V4.7_and_VAX_Workstation_Software_V3.3_on_SIMH) to get it running which works with the latest simh code and (theoretically) includes networking.
```shell
$ SHOW NET
VAX/VMS Network status for local node  1.13 GAIA on 16-JUN-1987 20:14:39.61

This is a nonrouting node, and does not have any network information.
The designated router for GAIA is node    0 GAIA.

$ SHOW DEV XQ /FULL

Device XQA0: is online, network device, device is a template only.

    Error count                    0    Operations completed                  0
    Owner process                 ""    Owner UIC                         [0,0]
    Owner process ID        00000000    Dev Prot                               
    Reference count                0    Default buffer size                 512

Device XQA2: is online, network device.

    Error count                    0    Operations completed                190
    Owner process           "NETACP"    Owner UIC                      [SYSTEM]
    Owner process ID        0000002F    Dev Prot                               
    Reference count                2    Default buffer size                1498
```

I say "theoretically" because I am not skilled enough in VMS 4.7.  If I got TELNET working, I could easily test DECLRP, so I'm looking into how people ever got [TCP/IP installed](https://supratim-sanyal.blogspot.com/2019/03/cmu-tcpip-vax-vms-4.7.html), but it remains a mystery. The best I can do is `SET HOST /DTE CSA0` while running simh reading and writing to named pipes, but that doesn't work with declrp.py.

**Comment by Asmodai** [comment]
simh's networking is pretty good if you use tap/tun rather than pcap.  For DECnet, I'll probably be going for VDE.  It's going to take slightly longer for me because I intend to use real hardware for it, so I can set up my DECserver 100 on the network segment.  Eventually, I'll track down a DELNI for good measure.

~~Have you seen https://www.openvmshobby.com/vax-vms/carnegie-mellon-tcpip-vax/ btw?~~  Just realised this is for OpenVMS 7.

That Discord with the 4.7 turnkey might be interesting, though:  https://cmg.telegraphics.net/classiccmp/vms/vms_tk.html is another source for a similar image.

edit: another solution would be to configure DECnet on VMS 4.7 and host an older version of Linux in a VM.  If it's sufficiently old enough, then that can run Linux-DECnet (kernel module for DECnet and https://sourceforge.net/projects/linux-decnet/files/dnprogs/2.61/ tools), then configure it as a bridge between the TCP/IP world and the DECnet world.  You'd probably want to use VDE or configure simh to use tun/tap, however.  pcap would probably be more trouble than its worth.

https://retrocmp.com/decnet/decnet-on-ubuntu-setup is useful for that.

**Comment by davidrg** [comment]
> That Discord

The [Digital Equipment Corporation Fan Spot](https://discord.gg/EgzAUJj2n) I assume. At least thats where I remember seeing that turnkey kit mentioned.

> edit: another solution would be to configure DECnet on VMS 4.7 and host an older version of Linux in a VM.

The LAT server that shipped with Linux-DECnet at least works fine on modern linux. LAT is its own layer 2 protocol with no real relationship to DECnet so it doesn't need any kernel modules or anything. Might be able to get latd running on a modern linux host, then do a `SET HOST/LAT` or whatever the command is to connect to the linux box from the emulated VAX.

**Comment by Asmodai** [comment]
True that, there's also support for it on FreeBSD these days too.

**Comment by hackerb9** [comment]
@j4james: 
I got some results from VWS VT200/ReGIS terminal emulator via [simh][] VAX
emulation. Of course, any results from Asmodai's genuine VAX would be more trustworthy.

I don't expect Python runs under VMS 4.7, so I pulled some
shenanigans. In brief: I ran the test script on the host computer, 
redirecting the filedescriptors to a TCP port exposed by SIMH to fake a
DZ11 serial device in VMS.

* [declrp.log.vms4.7][]
* [declrp.log.vms5.5][]

[declrp.log.vms4.7]: https://gist.githubusercontent.com/hackerb9/4b985e504911fc95207fdc59c5a3b399/raw/06738e450b88668e4fb53de786c6762249a818a2/declrp.log.vms4.7 "Locator testing script run using VWS ReGIS VT200 terminal under VMS 4.7"

[declrp.log.vms5.5]: https://gist.githubusercontent.com/hackerb9/4b985e504911fc95207fdc59c5a3b399/raw/06738e450b88668e4fb53de786c6762249a818a2/declrp.log.vms5.5  "Locator testing script run using VWS ReGIS VT200 terminal under VMS 4.7"

_____

I wrote up a detailed [howto][] for anyone who wants to do more testing with this terminal emulator. I'm certainly going to be experimenting with it to see how its ReGIS support compares to the VT340. In many ways it seems like a step backward, but it's a bit hard to tell as I'm not sure how to get SIMH to display in color.

[howto]: https://gist.github.com/hackerb9/4b985e504911fc95207fdc59c5a3b399 

[simh]: https://github.com/open-simh/simh "The Open SIMH project: A VAX emulator"

**Comment by Asmodai** [comment]
I don't actually own a real VAX... all my VAX stuff is done via simh.  Of course, this means now I should change that!  The ideal for VWS, I think, would be a VAXstation 3100 M38.  Those show up on eBay from time to time.

**Comment by j4james** [comment]
> I got some results from VWS VT200/ReGIS terminal emulator

@hackerb9 Thank you! It doesn't follow the standard as well as DECterm, but it's interesting to note some of the differences.

On the bad side:
* It doesn't report anything when the locator is disabled
* It doesn't repoort the correct position on double-width lines (unless maybe the terminal doesn't support DECDWL?)
* It doesn't take margins into account

On the good side:
* It does report the correct position in 132-column mode
* It does report a page number (although always 1, so assumedly doesn't support multiple pages)

On the weird side:
* Parameter values are zero padded!

And I realised now that the unexpected offsets seen for the `DECCOLM` test (in both terminals) can easily be explained if they're just switching the font to the best available size that approximates 132-column mode, without trying to maintain a fixed window size.

In DECterm it looks like 80-column mode is a 7x13 font, and 132-column mode is 5x12. In the VWS emulator it looks like 80-column mode is 8x15, and 132-column mode is 6x13. So in both case the font is narrower, but not enough to prevent the window growing in width. And because the new font is shorter, the window height assumedly decreases too. I think the VWS emulator may also be re-positioning the window to accommodate the width change.

**Comment by hackerb9** [comment]
Yup, the term in VWS jumps around as the window resizes for 132-column mode.
